File: client/java/client/src/main/java/dev/angzarr/client/SpeculativeClient.java
package dev.angzarr.client;

import dev.angzarr.*;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.StatusRuntimeException;

/**
 * Client for speculative (what-if) execution across coordinator types.
 *
 * <p>Speculative execution allows testing commands, events, and projections
 * without persisting results. Use this for:
 * <ul>
 *   <li>Form validation: "Will this order succeed?"</li>
 *   <li>Preview: "What events would this produce?"</li>
 *   <li>Testing: Verify business logic without polluting event store</li>
 * </ul>
 *
 * <p>Example:
 * <pre>{@code
 * var client = SpeculativeClient.connect("localhost:1310");
 * var response = client.aggregate(SpeculateAggregateRequest.newBuilder()
 *     .setCommand(command)
 *     .setEvents(priorEvents)
 *     .build());
 * }</pre>
 */
public class SpeculativeClient implements AutoCloseable {

    private final AggregateCoordinatorServiceGrpc.AggregateCoordinatorServiceBlockingStub aggregateStub;
    private final SagaCoordinatorServiceGrpc.SagaCoordinatorServiceBlockingStub sagaStub;
    private final ProjectorCoordinatorServiceGrpc.ProjectorCoordinatorServiceBlockingStub projectorStub;
    private final ProcessManagerCoordinatorServiceGrpc.ProcessManagerCoordinatorServiceBlockingStub pmStub;
    private final ManagedChannel channel;

    private SpeculativeClient(ManagedChannel channel) {
        this.channel = channel;
        this.aggregateStub = AggregateCoordinatorServiceGrpc.newBlockingStub(channel);
        this.sagaStub = SagaCoordinatorServiceGrpc.newBlockingStub(channel);
        this.projectorStub = ProjectorCoordinatorServiceGrpc.newBlockingStub(channel);
        this.pmStub = ProcessManagerCoordinatorServiceGrpc.newBlockingStub(channel);
    }

    /**
     * Connect to coordinator services at the given endpoint.
     *
     * @param endpoint the host:port to connect to
     * @return a connected SpeculativeClient
     * @throws Errors.ConnectionError if connection fails
     */
    public static SpeculativeClient connect(String endpoint) {
        try {
            String[] parts = endpoint.split(":");
            if (parts.length != 2) {
                throw new Errors.ConnectionError("Invalid endpoint format: " + endpoint);
            }
            ManagedChannel channel = ManagedChannelBuilder
                .forAddress(parts[0], Integer.parseInt(parts[1]))
                .usePlaintext()
                .build();
            return new SpeculativeClient(channel);
        } catch (NumberFormatException e) {
            throw new Errors.ConnectionError("Invalid port in endpoint: " + endpoint, e);
        }
    }

    /**
     * Connect using an endpoint from environment variable with fallback.
     *
     * @param envVar the environment variable name
     * @param defaultEndpoint the default endpoint if env var is not set
     * @return a connected SpeculativeClient
     */
    public static SpeculativeClient fromEnv(String envVar, String defaultEndpoint) {
        String endpoint = System.getenv(envVar);
        if (endpoint == null || endpoint.isEmpty()) {
            endpoint = defaultEndpoint;
        }
        return connect(endpoint);
    }

    /**
     * Create a client from an existing channel.
     *
     * @param channel the gRPC channel to use
     * @return a SpeculativeClient using the channel
     */
    public static SpeculativeClient fromChannel(ManagedChannel channel) {
        return new SpeculativeClient(channel);
    }

    /**
     * Execute a command speculatively against temporal aggregate state.
     *
     * @param request the speculative aggregate request
     * @return the command response (without persistence)
     * @throws Errors.GrpcError if the RPC fails
     */
    public CommandResponse aggregate(SpeculateAggregateRequest request) {
        try {
            return aggregateStub.handleSyncSpeculative(request);
        } catch (StatusRuntimeException e) {
            throw new Errors.GrpcError(e.getMessage(), e.getStatus().getCode());
        }
    }

    /**
     * Execute a projector speculatively against events.
     *
     * @param request the speculative projector request
     * @return the projection result (without persistence)
     * @throws Errors.GrpcError if the RPC fails
     */
    public dev.angzarr.Projection projector(SpeculateProjectorRequest request) {
        try {
            return projectorStub.handleSpeculative(request);
        } catch (StatusRuntimeException e) {
            throw new Errors.GrpcError(e.getMessage(), e.getStatus().getCode());
        }
    }

    /**
     * Execute a saga speculatively against events.
     *
     * @param request the speculative saga request
     * @return the saga response (without persistence)
     * @throws Errors.GrpcError if the RPC fails
     */
    public SagaResponse saga(SpeculateSagaRequest request) {
        try {
            return sagaStub.executeSpeculative(request);
        } catch (StatusRuntimeException e) {
            throw new Errors.GrpcError(e.getMessage(), e.getStatus().getCode());
        }
    }

    /**
     * Execute a process manager speculatively.
     *
     * @param request the speculative PM request
     * @return the PM response (without persistence)
     * @throws Errors.GrpcError if the RPC fails
     */
    public ProcessManagerHandleResponse processManager(SpeculatePmRequest request) {
        try {
            return pmStub.handleSpeculative(request);
        } catch (StatusRuntimeException e) {
            throw new Errors.GrpcError(e.getMessage(), e.getStatus().getCode());
        }
    }

    /**
     * Close the underlying channel.
     */
    @Override
    public void close() {
        if (channel != null && !channel.isShutdown()) {
            channel.shutdown();
        }
    }
}

File: client/csharp/Angzarr.Client/SpeculativeClient.cs
using Grpc.Core;
using Grpc.Net.Client;

namespace Angzarr.Client;

/// <summary>
/// Client for speculative (what-if) execution across coordinator types.
/// </summary>
/// <remarks>
/// Speculative execution allows testing commands, events, and projections
/// without persisting results. Use this for:
/// <list type="bullet">
///   <item>Form validation: "Will this order succeed?"</item>
///   <item>Preview: "What events would this produce?"</item>
///   <item>Testing: Verify business logic without polluting event store</item>
/// </list>
/// </remarks>
/// <example>
/// <code>
/// using var client = SpeculativeClient.Connect("http://localhost:1310");
/// var response = client.Aggregate(new SpeculateAggregateRequest
/// {
///     Command = command,
///     Events = priorEvents
/// });
/// </code>
/// </example>
public class SpeculativeClient : IDisposable
{
    private readonly AggregateCoordinatorService.AggregateCoordinatorServiceClient _aggregateStub;
    private readonly SagaCoordinatorService.SagaCoordinatorServiceClient _sagaStub;
    private readonly ProjectorCoordinatorService.ProjectorCoordinatorServiceClient _projectorStub;
    private readonly ProcessManagerCoordinatorService.ProcessManagerCoordinatorServiceClient _pmStub;
    private readonly GrpcChannel _channel;

    private SpeculativeClient(GrpcChannel channel)
    {
        _channel = channel;
        _aggregateStub = new AggregateCoordinatorService.AggregateCoordinatorServiceClient(channel);
        _sagaStub = new SagaCoordinatorService.SagaCoordinatorServiceClient(channel);
        _projectorStub = new ProjectorCoordinatorService.ProjectorCoordinatorServiceClient(channel);
        _pmStub = new ProcessManagerCoordinatorService.ProcessManagerCoordinatorServiceClient(channel);
    }

    /// <summary>
    /// Connect to coordinator services at the given endpoint.
    /// </summary>
    /// <param name="endpoint">The endpoint URL (e.g., "http://localhost:1310")</param>
    /// <returns>A connected SpeculativeClient</returns>
    /// <exception cref="ConnectionError">If connection fails</exception>
    public static SpeculativeClient Connect(string endpoint)
    {
        try
        {
            // Validate endpoint format
            if (!Uri.TryCreate(endpoint, UriKind.Absolute, out var uri))
            {
                // Try adding http:// prefix
                if (!Uri.TryCreate($"http://{endpoint}", UriKind.Absolute, out uri))
                {
                    throw new ConnectionError($"Invalid endpoint format: {endpoint}");
                }
            }

            // Validate it looks like a proper endpoint
            if (uri.Segments.Length > 1 || string.IsNullOrEmpty(uri.Host))
            {
                throw new ConnectionError($"Invalid endpoint format: {endpoint}");
            }

            var channel = GrpcChannel.ForAddress(uri);
            return new SpeculativeClient(channel);
        }
        catch (UriFormatException ex)
        {
            throw new ConnectionError($"Invalid endpoint format: {endpoint}", ex);
        }
    }

    /// <summary>
    /// Connect using an endpoint from environment variable with fallback.
    /// </summary>
    /// <param name="envVar">The environment variable name</param>
    /// <param name="defaultEndpoint">The default endpoint if env var is not set</param>
    /// <returns>A connected SpeculativeClient</returns>
    public static SpeculativeClient FromEnv(string envVar, string defaultEndpoint)
    {
        var endpoint = Environment.GetEnvironmentVariable(envVar);
        if (string.IsNullOrEmpty(endpoint))
        {
            endpoint = defaultEndpoint;
        }
        return Connect(endpoint);
    }

    /// <summary>
    /// Create a client from an existing channel.
    /// </summary>
    /// <param name="channel">The gRPC channel to use</param>
    /// <returns>A SpeculativeClient using the channel</returns>
    public static SpeculativeClient FromChannel(GrpcChannel channel)
    {
        return new SpeculativeClient(channel);
    }

    /// <summary>
    /// Execute a command speculatively against temporal aggregate state.
    /// </summary>
    /// <param name="request">The speculative aggregate request</param>
    /// <returns>The command response (without persistence)</returns>
    /// <exception cref="GrpcError">If the RPC fails</exception>
    public CommandResponse Aggregate(SpeculateAggregateRequest request)
    {
        try
        {
            return _aggregateStub.HandleSyncSpeculative(request);
        }
        catch (RpcException ex)
        {
            throw new GrpcError(ex.Message, ex.StatusCode);
        }
    }

    /// <summary>
    /// Execute a projector speculatively against events.
    /// </summary>
    /// <param name="request">The speculative projector request</param>
    /// <returns>The projection result (without persistence)</returns>
    /// <exception cref="GrpcError">If the RPC fails</exception>
    public Projection Projector(SpeculateProjectorRequest request)
    {
        try
        {
            return _projectorStub.HandleSpeculative(request);
        }
        catch (RpcException ex)
        {
            throw new GrpcError(ex.Message, ex.StatusCode);
        }
    }

    /// <summary>
    /// Execute a saga speculatively against events.
    /// </summary>
    /// <param name="request">The speculative saga request</param>
    /// <returns>The saga response (without persistence)</returns>
    /// <exception cref="GrpcError">If the RPC fails</exception>
    public SagaResponse Saga(SpeculateSagaRequest request)
    {
        try
        {
            return _sagaStub.ExecuteSpeculative(request);
        }
        catch (RpcException ex)
        {
            throw new GrpcError(ex.Message, ex.StatusCode);
        }
    }

    /// <summary>
    /// Execute a process manager speculatively.
    /// </summary>
    /// <param name="request">The speculative PM request</param>
    /// <returns>The PM response (without persistence)</returns>
    /// <exception cref="GrpcError">If the RPC fails</exception>
    public ProcessManagerHandleResponse ProcessManager(SpeculatePmRequest request)
    {
        try
        {
            return _pmStub.HandleSpeculative(request);
        }
        catch (RpcException ex)
        {
            throw new GrpcError(ex.Message, ex.StatusCode);
        }
    }

    /// <summary>
    /// Dispose the underlying channel.
    /// </summary>
    public void Dispose()
    {
        _channel?.Dispose();
    }
}
